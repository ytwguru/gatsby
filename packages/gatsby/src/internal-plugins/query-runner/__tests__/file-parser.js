jest.mock(`fs-extra`, () => {
  let mockFiles = {}
  return {
    __setMockFiles: newMockFiles => {
      mockFiles = Object.assign({}, newMockFiles)
    },
    readFile: (filePath, parser) =>
      new Promise(resolve => resolve(mockFiles[filePath])),
  }
})
jest.mock(`../../../utils/api-runner-node`, () => () => [])

const FileParser = require(`../file-parser`).default

describe(`File parser`, () => {
  const MOCK_FILE_INFO = {
    "no-query.js": `import React from "react"`,
    "page-query.js": `export const query = graphql\`
query PageQueryName {
  foo
}
\``,
    "page-query-no-name.js": `export const query = graphql\`
query {
  foo
}
\``,
    "static-query.js": `export default () => (
  <StaticQuery
    query={graphql\`query StaticQueryName { foo }\`}
    render={data => <div>{data.doo}</div>}
  />
)`,
    "static-query-no-name.js": `export default () => (
  <StaticQuery
    query={graphql\`{ foo }\`}
    render={data => <div>{data.foo}</div>}
  />
)`,
    "static-query-named-export.js": `export const Component = () => (
  <StaticQuery
    query={graphql\`query StaticQueryName { foo }\`}
    render={data => <div>{data.doo}</div>}
  />
)`,
    "static-query-closing-tag.js": `export default () => (
  <StaticQuery
    query={graphql\`{ foo }\`}
  >
    {data => <div>{data.foo}</div>}
  </StaticQuery>
)`,
    "page-query-and-static-query-named-export.js": `export const Component = () => (
  <StaticQuery
    query={graphql\`query StaticQueryName { foo }\`}
    render={data => <div>{data.doo}</div>}
  />
)
export const pageQuery = graphql\`query PageQueryName { foo }\`
`,
    "multiple-fragment-exports.js": `export const fragment1 = graphql\`
  fragment Fragment1 on RootQueryField {
    foo
  }
  fragment Fragment2 on RootQueryField {
    bar
  }
\`
export const fragment3 = graphql\`
  fragment Fragment3 on RootQueryField {
    baz
  }
\`
`,
  }

  const parser = new FileParser()

  beforeAll(() => {
    require(`fs-extra`).__setMockFiles(MOCK_FILE_INFO)
  })

  it(`extracts query AST correctly from files`, async () => {
    const results = await parser.parseFiles(Object.keys(MOCK_FILE_INFO))
    expect(results).toMatchSnapshot()
  })
})
