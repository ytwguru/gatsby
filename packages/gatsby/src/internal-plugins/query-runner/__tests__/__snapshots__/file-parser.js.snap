// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`File parser extracts query AST correctly from files 1`] = `
Map {
  "page-query.js" => Object {
    "definitions": Array [
      Object {
        "directives": Array [],
        "kind": "OperationDefinition",
        "loc": Object {
          "end": 30,
          "start": 1,
        },
        "name": Object {
          "kind": "Name",
          "loc": Object {
            "end": 20,
            "start": 7,
          },
          "value": "PageQueryName",
        },
        "operation": "query",
        "selectionSet": Object {
          "kind": "SelectionSet",
          "loc": Object {
            "end": 30,
            "start": 21,
          },
          "selections": Array [
            Object {
              "alias": undefined,
              "arguments": Array [],
              "directives": Array [],
              "kind": "Field",
              "loc": Object {
                "end": 28,
                "start": 25,
              },
              "name": Object {
                "kind": "Name",
                "loc": Object {
                  "end": 28,
                  "start": 25,
                },
                "value": "foo",
              },
              "selectionSet": undefined,
            },
          ],
        },
        "variableDefinitions": Array [],
      },
    ],
    "kind": "Document",
  },
  "page-query-no-name.js" => Object {
    "definitions": Array [
      Object {
        "directives": Array [],
        "kind": "OperationDefinition",
        "loc": Object {
          "end": 16,
          "start": 1,
        },
        "name": Object {
          "kind": "Name",
          "value": "pageQueryNoNameJs2538359280",
        },
        "operation": "query",
        "selectionSet": Object {
          "kind": "SelectionSet",
          "loc": Object {
            "end": 16,
            "start": 7,
          },
          "selections": Array [
            Object {
              "alias": undefined,
              "arguments": Array [],
              "directives": Array [],
              "kind": "Field",
              "loc": Object {
                "end": 14,
                "start": 11,
              },
              "name": Object {
                "kind": "Name",
                "loc": Object {
                  "end": 14,
                  "start": 11,
                },
                "value": "foo",
              },
              "selectionSet": undefined,
            },
          ],
        },
        "variableDefinitions": Array [],
      },
    ],
    "kind": "Document",
  },
  "static-query.js" => Object {
    "definitions": Array [
      Object {
        "directives": Array [],
        "hash": 407706182,
        "isStaticQuery": true,
        "kind": "OperationDefinition",
        "loc": Object {
          "end": 29,
          "start": 0,
        },
        "name": Object {
          "kind": "Name",
          "loc": Object {
            "end": 21,
            "start": 6,
          },
          "value": "StaticQueryName",
        },
        "operation": "query",
        "selectionSet": Object {
          "kind": "SelectionSet",
          "loc": Object {
            "end": 29,
            "start": 22,
          },
          "selections": Array [
            Object {
              "alias": undefined,
              "arguments": Array [],
              "directives": Array [],
              "kind": "Field",
              "loc": Object {
                "end": 27,
                "start": 24,
              },
              "name": Object {
                "kind": "Name",
                "loc": Object {
                  "end": 27,
                  "start": 24,
                },
                "value": "foo",
              },
              "selectionSet": undefined,
            },
          ],
        },
        "text": "query StaticQueryName { foo }",
        "variableDefinitions": Array [],
      },
    ],
    "kind": "Document",
  },
  "static-query-no-name.js" => Object {
    "definitions": Array [
      Object {
        "directives": Array [],
        "hash": 2605331189,
        "isStaticQuery": true,
        "kind": "OperationDefinition",
        "loc": Object {
          "end": 7,
          "start": 0,
        },
        "name": Object {
          "kind": "Name",
          "value": "staticQueryNoNameJs2605331189",
        },
        "operation": "query",
        "selectionSet": Object {
          "kind": "SelectionSet",
          "loc": Object {
            "end": 7,
            "start": 0,
          },
          "selections": Array [
            Object {
              "alias": undefined,
              "arguments": Array [],
              "directives": Array [],
              "kind": "Field",
              "loc": Object {
                "end": 5,
                "start": 2,
              },
              "name": Object {
                "kind": "Name",
                "loc": Object {
                  "end": 5,
                  "start": 2,
                },
                "value": "foo",
              },
              "selectionSet": undefined,
            },
          ],
        },
        "text": "{ foo }",
        "variableDefinitions": Array [],
      },
    ],
    "kind": "Document",
  },
  "static-query-named-export.js" => Object {
    "definitions": Array [
      Object {
        "directives": Array [],
        "hash": 407706182,
        "isStaticQuery": true,
        "kind": "OperationDefinition",
        "loc": Object {
          "end": 29,
          "start": 0,
        },
        "name": Object {
          "kind": "Name",
          "loc": Object {
            "end": 21,
            "start": 6,
          },
          "value": "StaticQueryName",
        },
        "operation": "query",
        "selectionSet": Object {
          "kind": "SelectionSet",
          "loc": Object {
            "end": 29,
            "start": 22,
          },
          "selections": Array [
            Object {
              "alias": undefined,
              "arguments": Array [],
              "directives": Array [],
              "kind": "Field",
              "loc": Object {
                "end": 27,
                "start": 24,
              },
              "name": Object {
                "kind": "Name",
                "loc": Object {
                  "end": 27,
                  "start": 24,
                },
                "value": "foo",
              },
              "selectionSet": undefined,
            },
          ],
        },
        "text": "query StaticQueryName { foo }",
        "variableDefinitions": Array [],
      },
    ],
    "kind": "Document",
  },
  "static-query-closing-tag.js" => Object {
    "definitions": Array [
      Object {
        "directives": Array [],
        "hash": 2605331189,
        "isStaticQuery": true,
        "kind": "OperationDefinition",
        "loc": Object {
          "end": 7,
          "start": 0,
        },
        "name": Object {
          "kind": "Name",
          "value": "staticQueryClosingTagJs2605331189",
        },
        "operation": "query",
        "selectionSet": Object {
          "kind": "SelectionSet",
          "loc": Object {
            "end": 7,
            "start": 0,
          },
          "selections": Array [
            Object {
              "alias": undefined,
              "arguments": Array [],
              "directives": Array [],
              "kind": "Field",
              "loc": Object {
                "end": 5,
                "start": 2,
              },
              "name": Object {
                "kind": "Name",
                "loc": Object {
                  "end": 5,
                  "start": 2,
                },
                "value": "foo",
              },
              "selectionSet": undefined,
            },
          ],
        },
        "text": "{ foo }",
        "variableDefinitions": Array [],
      },
    ],
    "kind": "Document",
  },
  "page-query-and-static-query-named-export.js" => Object {
    "definitions": Array [
      Object {
        "directives": Array [],
        "hash": 407706182,
        "isStaticQuery": true,
        "kind": "OperationDefinition",
        "loc": Object {
          "end": 29,
          "start": 0,
        },
        "name": Object {
          "kind": "Name",
          "loc": Object {
            "end": 21,
            "start": 6,
          },
          "value": "StaticQueryName",
        },
        "operation": "query",
        "selectionSet": Object {
          "kind": "SelectionSet",
          "loc": Object {
            "end": 29,
            "start": 22,
          },
          "selections": Array [
            Object {
              "alias": undefined,
              "arguments": Array [],
              "directives": Array [],
              "kind": "Field",
              "loc": Object {
                "end": 27,
                "start": 24,
              },
              "name": Object {
                "kind": "Name",
                "loc": Object {
                  "end": 27,
                  "start": 24,
                },
                "value": "foo",
              },
              "selectionSet": undefined,
            },
          ],
        },
        "text": "query StaticQueryName { foo }",
        "variableDefinitions": Array [],
      },
      Object {
        "directives": Array [],
        "kind": "OperationDefinition",
        "loc": Object {
          "end": 27,
          "start": 0,
        },
        "name": Object {
          "kind": "Name",
          "loc": Object {
            "end": 19,
            "start": 6,
          },
          "value": "PageQueryName",
        },
        "operation": "query",
        "selectionSet": Object {
          "kind": "SelectionSet",
          "loc": Object {
            "end": 27,
            "start": 20,
          },
          "selections": Array [
            Object {
              "alias": undefined,
              "arguments": Array [],
              "directives": Array [],
              "kind": "Field",
              "loc": Object {
                "end": 25,
                "start": 22,
              },
              "name": Object {
                "kind": "Name",
                "loc": Object {
                  "end": 25,
                  "start": 22,
                },
                "value": "foo",
              },
              "selectionSet": undefined,
            },
          ],
        },
        "variableDefinitions": Array [],
      },
    ],
    "kind": "Document",
  },
  "multiple-fragment-exports.js" => Object {
    "definitions": Array [
      Object {
        "directives": Array [],
        "kind": "FragmentDefinition",
        "loc": Object {
          "end": 53,
          "start": 3,
        },
        "name": Object {
          "kind": "Name",
          "loc": Object {
            "end": 21,
            "start": 12,
          },
          "value": "Fragment1",
        },
        "selectionSet": Object {
          "kind": "SelectionSet",
          "loc": Object {
            "end": 53,
            "start": 40,
          },
          "selections": Array [
            Object {
              "alias": undefined,
              "arguments": Array [],
              "directives": Array [],
              "kind": "Field",
              "loc": Object {
                "end": 49,
                "start": 46,
              },
              "name": Object {
                "kind": "Name",
                "loc": Object {
                  "end": 49,
                  "start": 46,
                },
                "value": "foo",
              },
              "selectionSet": undefined,
            },
          ],
        },
        "typeCondition": Object {
          "kind": "NamedType",
          "loc": Object {
            "end": 39,
            "start": 25,
          },
          "name": Object {
            "kind": "Name",
            "loc": Object {
              "end": 39,
              "start": 25,
            },
            "value": "RootQueryField",
          },
        },
      },
      Object {
        "directives": Array [],
        "kind": "FragmentDefinition",
        "loc": Object {
          "end": 106,
          "start": 56,
        },
        "name": Object {
          "kind": "Name",
          "loc": Object {
            "end": 74,
            "start": 65,
          },
          "value": "Fragment2",
        },
        "selectionSet": Object {
          "kind": "SelectionSet",
          "loc": Object {
            "end": 106,
            "start": 93,
          },
          "selections": Array [
            Object {
              "alias": undefined,
              "arguments": Array [],
              "directives": Array [],
              "kind": "Field",
              "loc": Object {
                "end": 102,
                "start": 99,
              },
              "name": Object {
                "kind": "Name",
                "loc": Object {
                  "end": 102,
                  "start": 99,
                },
                "value": "bar",
              },
              "selectionSet": undefined,
            },
          ],
        },
        "typeCondition": Object {
          "kind": "NamedType",
          "loc": Object {
            "end": 92,
            "start": 78,
          },
          "name": Object {
            "kind": "Name",
            "loc": Object {
              "end": 92,
              "start": 78,
            },
            "value": "RootQueryField",
          },
        },
      },
      Object {
        "directives": Array [],
        "kind": "FragmentDefinition",
        "loc": Object {
          "end": 53,
          "start": 3,
        },
        "name": Object {
          "kind": "Name",
          "loc": Object {
            "end": 21,
            "start": 12,
          },
          "value": "Fragment3",
        },
        "selectionSet": Object {
          "kind": "SelectionSet",
          "loc": Object {
            "end": 53,
            "start": 40,
          },
          "selections": Array [
            Object {
              "alias": undefined,
              "arguments": Array [],
              "directives": Array [],
              "kind": "Field",
              "loc": Object {
                "end": 49,
                "start": 46,
              },
              "name": Object {
                "kind": "Name",
                "loc": Object {
                  "end": 49,
                  "start": 46,
                },
                "value": "baz",
              },
              "selectionSet": undefined,
            },
          ],
        },
        "typeCondition": Object {
          "kind": "NamedType",
          "loc": Object {
            "end": 39,
            "start": 25,
          },
          "name": Object {
            "kind": "Name",
            "loc": Object {
              "end": 39,
              "start": 25,
            },
            "value": "RootQueryField",
          },
        },
      },
    ],
    "kind": "Document",
  },
}
`;
